### PR TITLE
Adds the Archon Projector, a new clockcult turret

### LINKED ISSUE #33797: Cogscarabs are no longer able to perform Script and Application tier scripture in noncult rounds.
Issue reported from Round ID: 80677 (/tg/Station Bagil)

It appears that Clockwork Scarabs are now unable to use anything above Driver tier scripture now during non-clockcult rounds. Cogscarabs are intended to have full access to all of the Scripture as I remember. I was able to perform scripture T2 and T3 about a week or so ago, so unless this was changed intentionally I don't know what could have broken it. The UI when trying to access these tiers looks like this:

dreamseeker_2017-12-23_14-36-44

### PR DESCRIPTION
**Pull request by ghost**
capture

MEET MR ARCHON PROJECTOR
HE SHOOTS HERETICS. BUT IF YOU GET IN THE WAY, HE MIGHT ACCIDENTALY SHOOT YOU TOO!

:cl: More Robust Than You
add: Clockwork Cultists can now build the Archon Projector, a new long range turret! It shoots projectiles, which may accidentally hit cultists in the crossfire, but do more damage than ocular wardens. The projectiles, however, lose damage if they pass through glass!
/:cl:

Anyways, the turret has a 1s cooldown between shots, consumes 175W power each shot, the projectile does 12.5 damage (damage is multiplied by 0.75 if it crosses a window tile)

Fixes #33797

### CODE REVIEW

**Comment by Armhulen** [comment]
BRAZEN BOLTER

**Comment by AdamElTablawy** [comment]
Stacking this and an ocular warden pretty much means people won't be able to evade the projectile, and seeing as these kinds of guns are usually at the front lines of defenses the 'it can hit servants too' isn't too much of a weakness.

Perhaps a constant power draw instead?

**Comment by MoreRobustThanYou** [comment]
How about it requires power to shoot?

File: code/game/gamemodes/clock_cult/clock_structures/archon_projector.dm [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+			P.fire()
+			last_fire = world.time
+	if(!target)
+		if(validtargets.len)
+			target = pick(validtargets)
+			playsound(src,'sound/machines/clockcult/ocularwarden-target.ogg',50,1)
+			visible_message("<span class='warning'>[src] whirrs, before its gun swivels to face [target]!</span>")
+			if(isliving(target))
+				var/mob/living/L = target
+				to_chat(L, "<span class='neovgre'><font size=2><i>\"TARGET ACQUIRED.\"</i></font></span>")
```
Comment (ShizCalev) [inline code/game/gamemodes/clock_cult/clock_structures/archon_projector.dm]: `<span class = 'neovgre_small'>\"Target`

File: code/game/gamemodes/clock_cult/clock_structures/archon_projector.dm [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+			else if(ismecha(target))
+				var/obj/mecha/M = target
+				to_chat(M.occupant, "<span class='neovgre'><font size=2><i>\"TARGET ACQUIRED.\"</i></font></span>")
```
Comment (ShizCalev) [inline code/game/gamemodes/clock_cult/clock_structures/archon_projector.dm]: `<span class = 'neovgre_small'>\"Target`

**Review by ShizCalev — CHANGES_REQUESTED** [review]

**Comment by Ratvar** [comment]
Make those manned, Robustin Gang Turret style?

**Comment by MoreRobustThanYou** [comment]
no

**Comment by MoreRobustThanYou** [comment]
don't give me flashbacks to gangmageddon

**Comment by Xhuis** [comment]
I can't say I'm a fan of this. It overlaps function-wise with the ocular warden.
**Ocular Warden:** Stationary defensive turret that automatically deals damage from range, hitscan, loses damage across distance and objects
**Archon Projector:** Stationary defensive turret that automatically deals damage from range, projectile, loses damage across windows
I would say that if you want this over Warden, to remove the Ocular Warden, or to replace this structure's functionality. The trap/trigger system was implemented to add different ways; I think this would be better if it was aimed manually and triggers would cause it to fire a projectile forwards with a cooldown between shots - essentially clockwork emitters or Brazen Bolters as they were going to be implemented before arm closed his PR.
Otherwise, this and the Warden are too similar to me and I feel it is superior in all situations due to higher DPS, more versatility, the ability to damage through obstacles, and familiarity. Looking at the code, it even looks like you just copypasted the warden's code and used it here. The same thing could be achieved with much less effort by making a lot of the warden's hardcoded things into variables, making the fire function into a proc, and making this a subtype that overrides that proc.

**Comment by MoreRobustThanYou** [comment]
@Xhuis You're forgetting about the metas here.

The current meta is to make wardens, then surrond them with 1 layer of windows. Bam, only lasers can kill the wardens now. At least the AP requires power to work. You can't even dodge a warden's gaze, and it has zero chance of friendly fire. I made the AP's bolt move faster than most projectiles, and it _**can**_ result in friendly fire. Wardens are ALOT more spammable than this (min 4 distance between APs, 3 between OWs), and more reliable (less chance of failure due to power, friendly fire, or simply missing)

**Comment by Xhuis** [comment]
That answers none of the questions I just posed. Why would you ever choose this over a warden?

**Comment by MoreRobustThanYou** [comment]
@Xhuis For example, with more complex base building, a Projector would be good at shooting down hallways. Or up close to the barrier, as when they thermite/deconstruct/bust down the walls that you have built on the Ark side of the barrier, chances are it'll be 1 wide, and it'll be rather hard to dodge the projectiles.  That's the one thing about APs vs OWs: Wardens are good against moving targets, while Projectors are better against non-moving targets or down narrow hallways

**Comment by Xhuis** [comment]
Hallways might be the only exception to the wardens' superiority here,due to the projector's extended sight range. Ocular wardens deal more damage per second overall, making them better against nonmoving targets. They can also be put behind girders and other dense structures to protect them from laser fire, but since projectors can not, they will be outclassed even in the ideal circumstance. Consider the following instance:

\# is a wall, . is a floor, @ is a crewmember, O is a warden, f is a projector, and | is a window.
```
# # # # # # # . #
# . . . . . . . | f
# @ # # # # # # #
```
Bob McLaser needs to traverse this hallway. He can't, however, without entering the projector's line of fire. The solution to this is for him to just move up one tile, fire a laser at the projector, and then go down a tile. This is repeated until the projector is gone, letting him move up. Now, imagine the same scenario with an ocular warden, but | is a *girder* instead:
```
# # # # # # # . #
# . . . . . . . | O
# @ # # # # # # #
```
Bob's strategy of laser-dodging no longer applies; he can approach within the hall because the warden can only see 3 tiles. However, he has to laser down an entire girder, which would take five lasers (girder = 100 integrity), plus an additional two to kill the warden itself (warden = 25 integrity). This is twice the amount needed to kill the projector in the same scenario, which only needs 3 because it has to be vulnerable to laser fire to damage enemies.

In this scenario, the warden would do the following damage:
`damage ratio = 1 - 0.1 (gear) - 0.15 (distance) = 0.75`
`15 (base dps) * 0.75 = 11.25 dps`
And the projector would do this much:
`12.5 (base dps) * 0.75 (crossed window) = 9.375 dps`
Not only is the warden more defensible and difficult to take down, *and* does more damage *despite being behind a tough girder*, but can be stacked more effectively than projectors. Make a minor alteration to the previous defense setup, and the strategy falls apart completely:
```
. . # O #
# # # | # # # . #
# . . . . . . . | O
# @ # # # # # # #
```
Because now, even if you get within safe firing range of the first warden, the second one can't be accessed without taking at least one damage tick from it, due to how fast its targeting processes, and you also have to either spend another six lasers on it to destroy it (while taking damage from it due to its positioning) or just let it live and chip the health of anyone going near it.

In both instances, projectors are both conceptually and mathematically less effective than wardens, and to boot, wardens are set-and-forget. Projectors are slow, vulnerable, and cost power to use. There is no reason in which a warden wouldn't perform its job better as it currently is.

**Comment by SgtxRootbeer** [comment]
i would put this at the front lines as undefended turrets, to abuse their longer range

**Comment by MoreRobustThanYou** [comment]
ok so i'll make the projector do more damage!

**Comment by Xhuis** [comment]
That still doesn't solve any of its problems! It still remains weak to laser fire because ducking continually behind cover nullifies any offensive ability it has. The only way to properly defend it and provide danger to anyone attempting to kill it is to use an ocular warden, and in that case, using an ocular warden instead of the projector is already more effective!

**Comment by MoreRobustThanYou** [comment]
Any way it could be fixed without completely redoing it?

I pulled this idea out of my ass: In addition to the projectile, it can connect a beam to a target, which will do very low burn damage and a moderate amount of stamina damage

**Comment by Xhuis** [comment]
Having a beam connected to targets in LoS would make it even *more* similar the ocular warden. In its current state (an automated defensive turret), in my opinion, it doesn't fit anywhere, because the ocular warden is already extensively cemented in that role. If it's to be a unique addition that other structures don't do better, then it should probably be reworked, though how it's reworked isn't my own choice.

**Comment by MoreRobustThanYou** [comment]
Idea #2: The projectile has a **_very weak_** homing capability, where is is capable of turning if the user thinks they can outsmart the projectile by moving 1 tile away

**Comment by Xhuis** [comment]
It seems impractical to me, since they can either just move two tiles away or duck behind cover. If you had a hallway, and to the left of the end of it a few tiles out was a projector, you could fire, then duck behind the wall. Unless the projectile circled the wall and chased the aggressor outright, the homing aspect could easily be avoided.

**Comment by MoreRobustThanYou** [comment]
Idea 3: the bolts can ricochet 2 times (but after the second ricochet, it starts to fade, becoming transparent and rapidly losing damage before qdeling about 5 seconds later)

**Comment by davethwave** [comment]
Just buff it a bit and make it a manned turret... make it a bit more unique than a normal tower defense turret and give it a bit a human interaction so that it can't be used in the same way as a warden as it needs a entrance/exit and so on and so forth.

**Comment by Xhuis** [comment]
It would be interesting, but I don't really get why it'd exist?

**Comment by Ratvar** [comment]
Robust was against manning due to "don't give me flashbacks to gangmageddon" already, duh

**Comment by davethwave** [comment]
oh, oops

**Comment by Armhulen** [comment]
Make them shoot in one direction, set by the cultists. make them shoot giant nails. then we'll have something unique

**Comment by KomradeSpectre** [comment]
Yes, but do that and you have walls of these put up where people are instantly dead the moment they walk out of the non-servant spawn.

**Comment by Xhuis** [comment]
If they're like emitters and only fire in one direction, you could only hit them with one projectile at a time, presumably.

**Comment by kevinz000** [comment]
dont forget to test

**Comment by MoreRobustThanYou** [comment]
@kevinz000 i tested it

**Comment by MoreRobustThanYou** [comment]
@Xhuis Can this be testmerged for balance testing and to see how people use it? Arm says you should OK it first.

File: code/game/gamemodes/clock_cult/clock_structures/archon_projector.dm [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+			else if(ismecha(target))
+				var/obj/mecha/M = target
+				to_chat(M.occupant, "<span class='neovgre_large'><i>\"TARGET ACQUIRED.\"</i></span>")
+		else if(prob(0.5)) //Extremely low chance because of how fast the subsystem it uses processes
+			if(prob(50))
+				visible_message("<span class='notice'>[src][pick(idle_messages)]</span>")
+			else
+				setDir(pick(GLOB.alldirs))//Random rotation
+
+/obj/structure/destructible/clockwork/archon_projector/proc/acquire_nearby_targets()
```
Comment (ChangelingRain) [inline code/game/gamemodes/clock_cult/clock_structures/archon_projector.dm]: this is entirely copied from ocular wardens; they should share this proc.

**Comment by Xhuis** [comment]
I guess I don't mind it being testmerged, but I really don't see it being worthwhile.

File: code/game/gamemodes/clock_cult/clock_structures/archon_projector.dm (hunk 3 above)
Comment (Armhulen) [reply code/game/gamemodes/clock_cult/clock_structures/archon_projector.dm]: hahaha

Comment (Armhulen) [reply code/game/gamemodes/clock_cult/clock_structures/archon_projector.dm]: you SUCK mister morty

Comment (Armhulen) [reply code/game/gamemodes/clock_cult/clock_structures/archon_projector.dm]: oh wait i do this too

Comment (MoreRobustThanYou) [reply code/game/gamemodes/clock_cult/clock_structures/archon_projector.dm]: hahahahahaha

Comment (MoreRobustThanYou) [reply code/game/gamemodes/clock_cult/clock_structures/archon_projector.dm]: when i'm not half-asleep i'll refactor this to a single proc

**Comment by MoreRobustThanYou** [comment]
Ok, I refactored ocular warden and the archon projector to be under a single type.

I buffed the damage of the AP to 17.5 (19.75 if the herald's beacon was activated). it should now properly pass grilles/tables/windows.

**Comment by Robustin** [comment]
Oh more more automatic Clockwork defenses.

/s

**Comment by optimumtact** [comment]
This doesn't seem to fit a niche

**Comment by MoreRobustThanYou** [comment]
speak english @optimumtact

**Comment by optimumtact** [comment]
add it on hippie and prove that it has value, otherwise you're just adding bloat.

**Comment by MoreRobustThanYou** [comment]
@optimumtact i don't contribute to hippie anymore, hippie is dying anyways.

**Pull request closed without merging**